### ISSUE TITLE
Problem with calculating Validation loss when packaging = False

### ISSUE THREAD

**Issue by szymonrucinski** [lvwerra/trl#478]
Whenever I am passing a 
```
 training_arguments = TrainingArguments(
        do_eval=True,
        evaluation_strategy="steps",
        eval_steps = 5,
        output_dir=f'./{MODEL_NAME}-finetuned',
        per_device_train_batch_size=config.per_device_train_batch_size,
        gradient_accumulation_steps=config.gradient_accumulation_steps,
        optim=config.optimizer,
        # logging_steps=config.eval_steps,
        learning_rate=config.learning_rate,
        fp16=False,
        num_train_epochs=config.max_epochs,
        max_grad_norm=config.max_gradient_norm,
        # group_by_length=True,
        lr_scheduler_type=config.lr_scheduler_type,
        load_best_model_at_end = True,
        report_to="wandb",)

 model_trainer = SFTTrainer(
            model,
            train_dataset=data["train"],
            eval_dataset=data["validation"],
            dataset_text_field="text",
            tokenizer = tokenizer
            max_seq_length=256,
            peft_config=peft_config,
            args=training_arguments,
            callbacks=[EarlyStoppingCallback(early_stopping_patience=3)]
        )
```
The training loss is being calculated but not a validation loss. I keep getting a no log every Nth step. This however changes if I pass `packaging=True to SFTTrainer`
In that case, I get both losses but I am unsure how packing text influences the quality of the training. 
How do you think I could fix it? 
Namely: How to properly process **data["validation"]** pass it and get validation loss properly calculated without packaging. Maybe that is some bug?

**Comment by dconathan** [comment]
I assume you mean `packing` and not `packaging`?

I am too seeing weird behavior here and not sure what is the expected behavior but yes I think it is a bug... I think this is related to the fact that `ConstantLengthDataset` is an iterable dataset instead of a sequence/map style dataset

Looking at this line in `transformers.Trainer.evaluation_loop`:
https://github.com/huggingface/transformers/blob/f4e4b4d0e2dc248433e808594f7595292037d891/src/transformers/trainer.py#L3124

In my debugging if I look at `len(dataloader)` at that point it is non-empty but if I go `list(enumerate(dataloader))` it evaluates to empty list, so no eval is being done there which probably explains what you are seeing.

**Comment by dconathan** [comment]
nvm I think my bug was because my eval dataset was too small and it was getting truncated to 0 examples... could that be happening to you too?
